File: code/native/swift/runtime/crypto.swift
import Foundation
import CryptoKit

enum crypto {
    private static func hex<D: Sequence>(_ bytes: D) -> String where D.Element == UInt8 {
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
    static func sha256(_ input: String) -> String { return hex(SHA256.hash(data: Data(input.utf8))) }
    static func sha512(_ input: String) -> String { return hex(SHA512.hash(data: Data(input.utf8))) }
    static func md5(_ input: String) -> String { return hex(Insecure.MD5.hash(data: Data(input.utf8))) }
    static func hmacSha256(_ key: String, _ data: String) -> String {
        let mac = HMAC<SHA256>.authenticationCode(for: Data(data.utf8), using: SymmetricKey(data: Data(key.utf8)))
        return hex(mac)
    }
    static func hmacSha512(_ key: String, _ data: String) -> String {
        let mac = HMAC<SHA512>.authenticationCode(for: Data(data.utf8), using: SymmetricKey(data: Data(key.utf8)))
        return hex(mac)
    }
    static func randomBytes(_ size: Int) -> String {
        var generator = SystemRandomNumberGenerator()
        var bytes = [UInt8]()
        for _ in 0..<size { bytes.append(UInt8.random(in: UInt8.min...UInt8.max, using: &generator)) }
        return hex(bytes)
    }
}

File: code/native/kotlin/runtime/crypto.kt
import java.security.MessageDigest
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

object crypto {
    private fun hex(bytes: ByteArray): String = bytes.joinToString("") { "%02x".format(it) }
    private fun digest(algorithm: String, input: String): String = hex(MessageDigest.getInstance(algorithm).digest(input.toByteArray(Charsets.UTF_8)))
    fun sha256(input: String): String = digest("SHA-256", input)
    fun sha512(input: String): String = digest("SHA-512", input)
    fun md5(input: String): String = digest("MD5", input)
    private fun mac(algorithm: String, key: String, data: String): String {
        val instance = Mac.getInstance(algorithm)
        instance.init(SecretKeySpec(key.toByteArray(Charsets.UTF_8), algorithm))
        return hex(instance.doFinal(data.toByteArray(Charsets.UTF_8)))
    }
    fun hmacSha256(key: String, data: String): String = mac("HmacSHA256", key, data)
    fun hmacSha512(key: String, data: String): String = mac("HmacSHA512", key, data)
    fun randomBytes(size: Long): String {
        val bytes = ByteArray(size.toInt())
        java.security.SecureRandom().nextBytes(bytes)
        return hex(bytes)
    }
}
